### DISCUSSION TITLE
Infer TypedDict as dict

### DISCUSSION THREAD

**Discussion by Alex23rodriguez** [microsoft/pyright#6043] [Q&A] [answered]
semi-related to issue #1727, but backwards. 
Is there any reason why a typed dict can't be infered as a dict? The variable is, after all, a dict!
For example:
```python
from typing import TypedDict

Book = TypedDict("Book", {"title": str, "author": str})

b1: Book = {
    "title": "The Hitchhiker's Guide to the Galaxy",
    "author": "Douglas Adams",
}

def takes_dict(d: dict):
    # does stuff generic to any dictionary
    pass

takes_dict(b1) # pyright error!
```

Why can't `b1` be infered as a dict? I would understand if it had to be casted to a dict with a specific signature, such as `dict[str, int]`, but in this case it's just `dict[Unknown, Unknown]` 

it feels really awkward to have to cast it to a dict, knowing it is already a dict:
```python
takes_dict(dict(b1))
```
Can this be done better? what is the reasoning behind it?

**Comment by erictraut** [comment, marked as answer]
This isn't an issue of inference. It's an issue of type compatibility. Unfortunately, a `TypedDict` is not "type consistent" with a `dict`. You can read about why this is not type safe in [PEP 589](https://peps.python.org/pep-0589/#type-consistency) which introduced `TypedDict` to the Python type system. The problem is that a `dict` is mutable, so if you pass a reference to a `TypedDict`, the receiver (who thinks they have a `dict`) could mutate it in ways that violate the type definition of the `TypedDict`. For example, they could delete keys that are required to be present on that `TypedDict`.

A `TypedDict` _is_ compatible with a `Mapping`, which is the immutable form of a `dict`. If you modify your code accordingly, this will type check without error:

```python
def takes_mapping(d: Mapping[str, Any]):
    pass

takes_mapping(b1)
```

**Reply by Alex23rodriguez** [reply]
I see. I knew I must have been missing something. Thanks for the answer!
